From 3c031744a660828c97f1d56f90e9ad495139d1e1 Mon Sep 17 00:00:00 2001
From: Christophe Guibout <christophe.guibout@st.com>
Date: Wed, 6 Apr 2022 18:23:18 +0200
Subject: [PATCH 1/2] plat/st: add management of boot partition

Give next stage boot the boot mode used (0 for normal boot,
2 for recovery).

Change-Id: I0cc406bfcb1f5a631733cf2eba6a3d9fe3400795
---
 plat/st/common/bl2_io_storage.c             |  1 +
 plat/st/stm32mp1/include/stm32mp1_context.h |  6 ++++++
 plat/st/stm32mp1/stm32mp1_private.c         | 17 +++++++++++++++++
 3 files changed, 24 insertions(+)

diff --git a/plat/st/common/bl2_io_storage.c b/plat/st/common/bl2_io_storage.c
index 2d2150cf4..f6c9ef2ad 100644
--- a/plat/st/common/bl2_io_storage.c
+++ b/plat/st/common/bl2_io_storage.c
@@ -641,6 +641,7 @@ void plat_fwu_set_images_source(const struct fwu_metadata *metadata)
 	io_block_spec_t *image_spec;
 
 	boot_idx = plat_fwu_get_boot_idx();
+	stm32_save_boot_mode((boot_idx == 1U) ? BOOT_RECOVERY : BOOT_NORMAL);
 	assert(boot_idx < NR_OF_FW_BANKS);
 
 	for (i = 0U; i < NR_OF_IMAGES_IN_FW_BANK; i++) {
diff --git a/plat/st/stm32mp1/include/stm32mp1_context.h b/plat/st/stm32mp1/include/stm32mp1_context.h
index 703fdce37..5d5d35e03 100644
--- a/plat/st/stm32mp1/include/stm32mp1_context.h
+++ b/plat/st/stm32mp1/include/stm32mp1_context.h
@@ -12,6 +12,11 @@
 
 #include <drivers/st/stm32_rtc.h>
 
+enum boot_mode {
+	BOOT_NORMAL = 0x00,
+	BOOT_RECOVERY = 0x02,
+};
+
 void stm32_clean_context(void);
 int stm32_save_context(uint32_t zq0cr0_zdata,
 		       struct stm32_rtc_calendar *rtc_time,
@@ -26,6 +31,7 @@ uint32_t stm32_get_zdata_from_context(void);
 int stm32_get_pll1_settings_from_context(void);
 bool stm32_are_pll1_settings_valid_in_context(void);
 bool stm32_pm_context_is_valid(void);
+int stm32_save_boot_mode(enum boot_mode mode);
 void stm32_save_ddr_training_area(void);
 void stm32_restore_ddr_training_area(void);
 uint32_t stm32_pm_get_optee_ep(void);
diff --git a/plat/st/stm32mp1/stm32mp1_private.c b/plat/st/stm32mp1/stm32mp1_private.c
index 1e0dda767..6f5a2edcd 100644
--- a/plat/st/stm32mp1/stm32mp1_private.c
+++ b/plat/st/stm32mp1/stm32mp1_private.c
@@ -27,6 +27,8 @@
 #define BOARD_ID_VARCPN_SHIFT		12
 #define BOARD_ID_REVISION_MASK		GENMASK(11, 8)
 #define BOARD_ID_REVISION_SHIFT		8
+#define TAMP_BOOT_MODE_MASK		GENMASK(7, 0)
+#define TAMP_BOOT_MODE_SHIFT		0
 #define BOARD_ID_VARFG_MASK		GENMASK(7, 4)
 #define BOARD_ID_VARFG_SHIFT		4
 #define BOARD_ID_BOM_MASK		GENMASK(3, 0)
@@ -817,6 +819,21 @@ bool stm32mp1_is_wakeup_from_standby(void)
 	return stm32_pm_context_is_valid();
 }
 
+int stm32_save_boot_mode(enum boot_mode mode)
+{
+	uint32_t bkpr_itf_idx = tamp_bkpr(TAMP_BOOT_ITF_BACKUP_REG_ID);
+
+	clk_enable(RTCAPB);
+
+	mmio_clrsetbits_32(bkpr_itf_idx,
+			   TAMP_BOOT_MODE_MASK,
+			   mode << TAMP_BOOT_MODE_SHIFT);
+
+	clk_disable(RTCAPB);
+
+	return 0;
+}
+
 void stm32_save_boot_interface(uint32_t interface, uint32_t instance)
 {
 	uintptr_t bkpr_itf_idx = tamp_bkpr(TAMP_BOOT_ITF_BACKUP_REG_ID);
-- 
2.25.1

